server.go

```diff
@@ -1856,6 +1856,57 @@ func (s *Server) Shutdown() error {
 	return nil
 }
 
+// ShutdownWithContext support shutdown with timeout and timeout happens after all listener already close.
+func (s *Server) ShutdownWithContext(ctx context.Context) (err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	atomic.StoreInt32(&s.stop, 1)
+	defer atomic.StoreInt32(&s.stop, 0)
+
+	if s.ln == nil {
+		return nil
+	}
+
+	for _, ln := range s.ln {
+		if err = ln.Close(); err != nil {
+			return err
+		}
+	}
+
+	if s.done != nil {
+		close(s.done)
+	}
+
+	// Closing the listener will make Serve() call Stop on the worker pool.
+	// Setting .stop to 1 will make serveConn() break out of its loop.
+	// Now we just have to wait until all workers are done or timeout.
+	ticker := time.NewTicker(time.Millisecond * 100)
+	defer ticker.Stop()
+END:
+	for {
+		s.closeIdleConns()
+
+		if open := atomic.LoadInt32(&s.open); open == 0 {
+			break
+		}
+		// This is not an optimal solution but using a sync.WaitGroup
+		// here causes data races as it's hard to prevent Add() to be called
+		// while Wait() is waiting.
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+			break END
```

Comment:
I'm wondering, should this then force close all connections? What does `net/http` do in this case?

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/net/http/server.go;l=2731

Reply:
> I'm wondering, should this then force close all connections? What does `net/http` do in this case?

The use of `shutdown` does not force all connections to close, but rather uses context to control the timeout after all `listener` have been closed.